src/states/elder/mod.rs

```diff
@@ -1737,6 +1689,40 @@ impl Approved for Elder {
 
         Ok(())
     }
+
+    // Checking members vote status and vote to remove those non-resposive nodes.
+    #[cfg(feature = "mock_parsec")]
+    fn check_voting_status(&mut self) {
+        let unresponsive_nodes = self.chain.check_vote_status();
+        for pub_id in unresponsive_nodes.iter() {
+            self.vote_for_event(AccumulatingEvent::Offline(*pub_id));
+        }
+    }
+
+    #[cfg(not(feature = "mock_parsec"))]
```

Comment:
Why can't we use this impl for `feature = "mock_parsec"` too?

Reply:
When using mock_parsec, the polling will give all peers' votes in one polling loop.
Which means there will be less extra polling that will make the unresponsive nodes to be accused for enough times.

Reply:
So is this just an optimization to make `mock_parsec` tests faster?

Reply:
no. more about avoiding the edge case that a disconnected node not being accused, due to our polling stopped too quick.

Reply:
Can that happen though? The polling will only stop when there are no unpolled observations. So after the polling stops, everyone will have seen all the injected user-data votes and they will also see that the dropped node voted for none of them. So they should always accuse it. Or am I missing something?

Reply:
if want to use the same code as to the mock_base, the diff of mock_parsec here is that all observations comes in within one polling loop. which means there is no `unpolled observations` left over and the polling could be stopped soon.
Meanwhile we are still expecting the dropped node to be accused for couple of more times.

Reply:
Parsec poll happens only when receiving gossip. So right after you inject all those fake votes, `has_unpolled_observations` will return `true` for everyone, because everyone put some observations into their parsecs, but no one polled them yet. So when `poll_and_resend` is called, it will do at least one iteration, so at least some nodes will receive gossips. Because of how mock parsec is implemented, that would cause those nodes to reach consensus on all the votes (this would not be the case in real parsec). So that means they will also see that the dropped node did not vote and will accuse it. There might be some nodes that did not receive any gossip, but that means they did not poll their parsecs so they will still return `true` from `has_unpolled_observations`, which will trigger another iteration of polling. This will go on until everyone received at least one gossip which mean they also reached consensus on all the votes so they will all accuse the dropped node.

So I still don't see how in mock_parsec we can miss accusing the node.

Reply:
The issue only happens for mock_base, not for mock_parsec.
mock_parsec is totally fine, that's why its function doesn't need extra logic handling (looking for extra accusations).

However, if want to use the same code as to mock_parsec, then the problem comes, because there probably won't be enough extra accusations.

Reply:
>  there probably won't be enough extra accusations

Ah OK, I think I see the problem now.

Reply:
So my reasoning for starting this thread is that I think it's worth to avoid conditional compilation as much as possible, because it can cause issues like tests passing with mock parsec but failing with real parsec (or vice-versa). I'm OK with doing it this way though, if there is no other way of it would be too much hassle. I have alternative idea though: in `verify_dropped_nodes`, we can use the `continue_if` predicate (by using `poll_and_resend_with_options`) to poll until the nodes are voted offline by everyone. A downside to that is that the tests might be a bit slower. 

Not sure what's best. What do the other reviewers think?

Reply:
using `poll_and_resend_with_options` won't help here for mock_base, as the issue is there is valid node being accused.  :(

>  it can cause issues like tests passing with mock parsec but failing with real parsec (or vice-versa)

Well, the mock_parsec and real_parsec already has different behaviour here (one gives everything within one poll loop, the other gives things gradually). Hence we have to have different code to handle diff situation.  :(

Reply:
Valid node being accused is solved by the multiple accusation thing you have in this PR. All I'm saying is to use that same logic for `mock_parsec` too, where the `poll_and_resend_with_options` could be used to prevent the polling to stop to early.

But you are right, there logic between real parsec and mock parsec is already different enough, so having a single implementation for this wouldn't achieve much. So I think I'm OK with this as long as the other reviewers are. @jeanphilippeD , @octol - what do you think guys?

Reply:
could be a good idea to use `poll_and_resend_with_options `, will have a try

Reply:
Tricky one

> I have alternative idea though: in verify_dropped_nodes, we can use the continue_if predicate > (by using poll_and_resend_with_options) to poll until the nodes are voted offline by everyone. A > downside to that is that the tests might be a bit slower.

If it was my choice and it was feasible I'd opt for slightly slower and avoid the conditional compilation. If it works that is...

Reply:
@madadam , tried with `poll_and_resend_with_options `, however it doesn't work. 
As `approved::parsec_poll` only got called when there is a parsec gossip, which only created when gossip_timeout got ticked. However, the advancing of clock only happens when `poll_all` or `is_busy` returns true within `poll_and_resend_with_options `. Unfortunately, they are both `false` in this case. Which means no matter how many times polled, there is no clock advancing , hence no gossip created.

I am not sure if it worth to change the behaviour of `poll_and_resend_with_options ` ?

Reply:
Hm, interesting and good catch. 

> I am not sure if it worth to change the behaviour of poll_and_resend_with_options 

I think it is worth it, because without it it's not as useful. But doesn't have to be in this PR. Up to you.

Reply:
changing the behaviour fails other test :(
I'd prefer leave it in a separate PR

Reply:
OK, then let's leave it as is.